keps/sig-auth/2862-fine-grained-kubelet-authz/README.md

```diff
@@ -0,0 +1,1013 @@
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Since we are adding a second SubjectAccessReview request the latency for some
+requests will increase. However, `kubelet` has a SubjectAccessReview response
+cache, so subsequent requests should result in cache hits.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+To determine if the caller has the required permissions for a particular request
+kubelet creates a SubjectAccessReview request to `kube-apiserver`. The
+SubjectAccessReviews are currently populated with:
+
+```yaml
+apiVersion: authorization.k8s.io/v1
+kind: SubjectAccessReview
+spec:
+  user: user1
+  uid: 64167384-10aa-4361-bcef-526ab51d1e4d
+  groups:
+  - groups1
+  resourceAttributes:
+    group: ""
+    version: "v1"
+    resource: "nodes"
+    namespace: ""
+    name: "node-1"
+    subresource: "proxy"
+    verb: "GET"
+```
+
+The following information is passed into the SubjectAccessReview request
+- Requesting user (generally from a TokenReviewRequest or client certificate authentication)
+- Request verb
+- A resource request with:
+  - APIGroup: ""
+  - APIVersion: "v1"
+  - Resource: "nodes"
+  - Name: nodeName (the name of the node the request was against)
+  - Subresource: which, before this KEP, can be one of: `proxy`, `log`, `metrics`, `spec`, `stats`
+
+The subresource is determined by the path of the request. `kube-apiserver` upon
+receiving this request will check if the user in the SubjectAccessReview is authorized
+to access the relevant resource or subresource. For example, if a cluster is using RBAC
+then this ClusterRole and ClusterRoleBinding might allow access:
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: nodes-proxy
+rules:
+- apiGroups: [""]
+  resources: ["nodes/proxy"]
+  verbs: ["get", "create"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: nodes-proxy-global
+subjects:
+- apiGroup: rbac.authorization.k8s.io
+  kind: User
+  name: user1
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: nodes-proxy
+```
+
+We will add a new feature-gate `KubeletFineGrainedAuthz` that will be defaulted to
+`false` in `alpha`. When this feature-gate is set to `true` kubelet will first send
+a SubjectAccessReview specifically for the `healthz` or `pods` endpoint, based on the path of the
+request made to the kubelet.
+If that request fails kubelet will retry with the coarse-grained verb (`proxy`).
```

Comment:
Add a note about the cache here. I think the Kubelet should check the cache fore both resources before sending either request.

Another optimization to consider: if there is an expired cache entry for the proxy subresource, check that subresource first.

Reply:
Added a note about caching. I don't think the webhook authorizer which implements the cache allows that level of control over the requests. We may end up sending a request that would fail if the TTL of the deny cache is over. However the cache TTLs are pretty forgiving and this will only happen for requests that are infrequent. I want to avoid adding another layer of caching to keep the changes small, but we could add caching in the new Authorizer layer which calls into the webhook authorizer. WDYT?

Reply:
Do we cache denials? If so, I think it's OK to leave it out, we can always optimize later. I might have been misremembering, thinking we don't cache denials...

Reply:
Yeah the webhook authorizer which is what kubelet uses and we won't change that does have cache for denials. https://github.com/kubernetes/kubernetes/blob/0aea469b96a1c0fed48dd2d4a0af96d36e24968c/staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go#L280C6-L280C17